### ISSUE TITLE
Torchvision import breaks OpenCV code

### ISSUE THREAD

**Issue by Whiax** [pytorch/vision#3856]
## 🐛 Bug

When I import torchvisions in an opencv python code, the code crashes with "Segmentation Fault" (in console) or "Restarting kernel" in Spyder / iPython.

## To Reproduce

1. Execute the code, the execution ends
2. Execute the code and uncomment the torchvision import, the execution gives "Segmentation fault"
3. Changing the order of import (torchvision then cv2) removes the issue for this short code sample but not for my complete code

```
import cv2
#import torchvision

pth = 'video.mp4'
cap = cv2.VideoCapture(pth)

ret,frame = cap.read()
cv2.imshow('frame',frame)
```

I also have an environment using Pytorch 1.5, the same code sample works fine under this environment.
Thanks for checking if this is a real issue

## Environment

PyTorch version: 1.8.0
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.10

Python version: 3.7 (64-bit runtime)
Python platform: Linux-5.4.0-65-generic-x86_64-with-debian-bullseye-sid
Is CUDA available: True
CUDA runtime version: Could not collect

cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] numpydoc==1.1.0
[pip3] torch==1.8.0
[pip3] torchaudio==0.8.0a0+a751e1d
[pip3] torchfile==0.1.0
[pip3] torchlars==0.1.2
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.9.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.1.1               h6406543_8    conda-forge
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py37he8ac12f_0  
[conda] mkl_fft                   1.2.0            py37h23d657b_0  
[conda] mkl_random                1.1.1            py37h0573a6f_0  
[conda] numpy                     1.19.2           py37h54aff64_0  
[conda] numpy-base                1.19.2           py37hfa32c7d_0  
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] pytorch                   1.8.0           py3.7_cuda11.1_cudnn8.0.5_0    pytorch
[conda] torch                     1.7.1                    pypi_0    pypi
[conda] torchaudio                0.7.2                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchlars                 0.1.2                     dev_0    <develop>
[conda] torchvision               0.9.0                py37_cu111    pytorch

**Comment by fmassa** [comment]
Thanks a lot for the report!

I believe this might be due to version mismatches between either libjpeg / libpng or ffmpeg.

If you are on a linux machine, can do you run your script on gdb to get the rest of the stack trace? Something like
```
gdb python
```
and then inside gdb
```
run my_script.py
```
and when it crashes run
```
bt
```
and paste the output here?

cc @andfoy as this is probably related to version mismatches of one of the IO functions

**Comment by Whiax** [comment]
Yep I can do that. The output gives:

> bt
> [New Thread 0x7fffb5993700 (LWP 3854944)]
> [New Thread 0x7fffba194700 (LWP 3854945)]
> [New Thread 0x7fffba995700 (LWP 3854946)]
> [New Thread 0x7fffbd196700 (LWP 3854947)]
> [New Thread 0x7ffe872d1700 (LWP 3854948)]
> [New Thread 0x7ffe86ad0700 (LWP 3854949)]
> [New Thread 0x7ffe862cf700 (LWP 3854950)]
> [New Thread 0x7ffe85ace700 (LWP 3854951)]
> [New Thread 0x7ffe852cd700 (LWP 3854952)]
> [New Thread 0x7ffe84acc700 (LWP 3854953)]
> [New Thread 0x7ffe842cb700 (LWP 3854954)]
> [New Thread 0x7ffe83aca700 (LWP 3854955)]
> [New Thread 0x7ffe832c9700 (LWP 3854956)]
> [New Thread 0x7ffe82ac8700 (LWP 3854957)]
> [New Thread 0x7ffe822c7700 (LWP 3854958)]
> [New Thread 0x7ffe81ac6700 (LWP 3854959)]
> [New Thread 0x7ffe812c5700 (LWP 3854960)]
> [New Thread 0x7ffe80ac4700 (LWP 3854961)]
> [New Thread 0x7ffe802c3700 (LWP 3854962)]
> [New Thread 0x7ffe7fac2700 (LWP 3854963)]
> [New Thread 0x7ffe7f2c1700 (LWP 3854964)]
> [New Thread 0x7ffe7eac0700 (LWP 3854965)]
> [New Thread 0x7ffe7e2bf700 (LWP 3854966)]
> [New Thread 0x7ffe7dabe700 (LWP 3854967)]
> --Type <RET> for more, q to quit, c to continue without paging--bt
> 
> Thread 1 "python" received signal SIGSEGV, Segmentation fault.
> 0x00007fffc3a563e6 in ?? ()
>    from /home/x/anaconda3/lib/python3.7/site-packages/av/../av.libs/libxcb-77222338.so.1.1.0
> 

It seems that the problem is linked to the X Window System maybe? I'm using Ubuntu + XFCE (Xubuntu). I guess cv2 calls X when it wants to create a window on my computer and somehow torchvision interfers with that maybe...

**Comment by fmassa** [comment]
the segfault happens in `from /home/x/anaconda3/lib/python3.7/site-packages/av/../av.libs/libxcb-77222338.so.1.1.0`, so I would suspect this is related to pyav and not torchvision per se.

Can you try running your demo code with importing av instead of torchvision and see if you still face the segfault? Something like
```python
import av
import cv2

pth = 'video.mp4'
cap = cv2.VideoCapture(pth)

ret,frame = cap.read()
cv2.imshow('frame',frame)
```

This will help us know where the issue is

**Comment by Whiax** [comment]
The code works with "import av" in place of "import torchvision". I can try other imports if you have a list.

**Comment by fmassa** [comment]
Thanks for the info. I'll let @andfoy follow-up on this with you

**Comment by andfoy** [comment]
Have you tried executing your code outside IPython in Spyder? We've recieved similar issues related to segfaults with cv2 in Spyder

**Comment by Whiax** [comment]
Hi, yes it's not specific to iPython. The segmentation fault is when I directly use python in the usual console.
It's quite hard for me to understand this issue. In another more complex code where there are imports of cv2 and torchvision, it crashes at a line where I call plt.plot(x) in command line (same gdb trace to pyav file). In spyder, it doesn't crash. And if I change the matplotlib backend from qt4/5 to tkagg, it also doesn't crash in console.

**Comment by Whiax** [comment]
The difference between the two environments:

```
> av                        8.0.3                    pypi_0    pypi
> blis                      0.7.4                    pypi_0    pypi
< ca-certificates           2021.1.19            h06a4308_0  
> ca-certificates           2020.12.5            ha878542_0    conda-forge
< certifi                   2020.12.5        py37h06a4308_0  
> catalogue                 1.0.0                    pypi_0    pypi
> certifi                   2020.12.5        py37h89c1867_1    conda-forge
> conda                     4.9.2            py37h89c1867_0    conda-forge
> conda-build               3.21.4           py37h06a4308_0  
> conda-env                 2.6.0                         1  
< cudatoolkit               10.1.243             h6bb024c_0  
> cudatoolkit               11.1.1               h6406543_8    conda-forge
> cymem                     2.0.5                    pypi_0    pypi
> ffmpeg                    4.3                  hf484d3e_0    pytorch
> gnutls                    3.6.13               h85f3911_1    conda-forge
> inflect                   5.2.0                    pypi_0    pypi
> lame                      3.100             h7f98852_1001    conda-forge
< _libgcc_mutex             0.1                        main  
< libgcc-ng                 9.1.0                hdf63c60_0  
> _libgcc_mutex             0.1                 conda_forge    conda-forge
> libgcc-ng                 9.3.0               h2828fa1_18    conda-forge
> libgomp                   9.3.0               h2828fa1_18    conda-forge
> libiconv                  1.16                 h516909a_0    conda-forge
< libstdcxx-ng              9.1.0                hdf63c60_0  
> libstdcxx-ng              9.3.0               h6de172a_18    conda-forge
> lpips                     0.1.3                    pypi_0    pypi
> murmurhash                1.0.5                    pypi_0    pypi
> nettle                    3.6                  he412f7d_0    conda-forge
> openh264                  2.1.1                h780b84a_0    conda-forge
> _openmp_mutex             4.5                       1_gnu    conda-forge
< openssl                   1.1.1i               h27cfd23_0  
> opensimplex               0.3                      pypi_0    pypi
> openssl                   1.1.1j               h7f98852_0    conda-forge
> plac                      1.1.3                    pypi_0    pypi
> preshed                   3.0.5                    pypi_0    pypi
< pytorch                   1.5.1           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
> pytorch                   1.8.0           py3.7_cuda11.1_cudnn8.0.5_0    pytorch
> spacy                     2.3.5                    pypi_0    pypi
> spacy-udpipe              0.3.2                    pypi_0    pypi
> srsly                     1.0.5                    pypi_0    pypi
> thinc                     7.4.5                    pypi_0    pypi
< torch                     1.4.0                    pypi_0    pypi
> torch                     1.7.1                    pypi_0    pypi
< torchvision               0.6.1                py37_cu101    pytorch
> torchvision               0.9.0                py37_cu111    pytorch
> ufal-udpipe               1.2.0.3                  pypi_0    pypi
> unidecode                 1.2.0                    pypi_0    pypi
> wasabi                    0.8.2                    pypi_0    pypi
```

**Comment by andfoy** [comment]
Have you tried creating a clean environment that uses just conda-forge/defaults and not pip packages? That combination is always guaranteed to generate segfaults

**Comment by andfoy** [comment]
Also, you have multiple versions of torchvision and pytorch installed

**Comment by andfoy** [comment]
(This is completely out of scope) Next time, when you use Spyder alongside binary libraries, we recommend you to set a separate environment for running Spyder and other to install your libraries, by using Spyder-kernels:  https://docs.spyder-ide.org/5/faq.html?highlight=environment#using-existing-environment

**Comment by Whiax** [comment]
Yeah I saw the multiple versions, not sure if that could be the issue. 
I'm recreating a new clean environment to check that.

> conda install torchvision -c pytorch
> conda install -c conda-forge opencv

If I do that, pytorch goes from 1.8.1 to 1.8.0 and torchvision also from 0.9.1 to 0.9.0. It installs opencv 4.5.2. 
And yeah ok, doing that **solves the issue**, so I guess it was on my side.
Sorry for the trouble, it's hard to get from where the bugs come from with these weird behaviors.
_

For those with the same issue:

- **Create a clean your conda environment**

Thanks for the help!

**Comment by CristoJV** [comment]
I can confirm the issue; it occurs for me as well with torchvision==0.16.0+cu118 and opencv_python==4.9.0.

```python
import cv2
import torchvision
if __name__ == "__main__":
    frame = cv2.imread("/app/tests/res/smiling_face.png")
    # Display the image
    cv2.imshow("Smiling face", frame)
    # Wait for a key press
    cv2.waitKey(0)
    
```
Version Compatibility Issue:
With torchvision==0.15.2+cu118, the function cv2.imshow() does not respond (freezes) when used alongside opencv==4.9.0.
To display images using OpenCV, it's necessary to comment out the torchvision import.

**Comment by TheotimeLH** [comment]
Happened to me today, but with pytorch directly (not torchvision).
To change the version of another package, I created a fresh env with both torch and opencv. Few hours later, I realized that under some conditions my code freeze; in which case I couldn't even stop it using Ctrl-C (on Linux) (in Jupyter, if I try to stop the kernel, Jupyter crashes). 

The freeze happens during loss.backward(); but as I can't use Ctrl-C, I cannot get a good Exception to see where it is stuck. 
Apparently it happens when my model uses a DDPMScheduler from https://huggingface.co/docs/diffusers/en/installation 

But for sure it's related to this issue; because after a few hours, I realized it only happens when opencv is imported somewhere.

To sum up; some torch models (used in Diffusers) + Opencv => freeze during loss.backward()
I tried both torch 2.1.2, and the latest 2.5.1;
Both Opencv 4.10.0 and 4.9.0
